### PR TITLE
[cudax] Simplify implementation of device attributes

### PR DESCRIPTION
**Pull request by davebayer**
This PR simplifies implementation of device attributes and removes duplicate code.

### CODE REVIEW

**Review by miscco — APPROVED** [review]

File: cudax/include/cuda/experimental/__device/attributes.cuh [hunk 1]
```diff
@@ -32,31 +32,8 @@ namespace cuda::experimental
...
 
-_CCCL_NODISCARD inline int __get_attr_impl(::cudaDeviceAttr __attr, int __dev)
-{
-  int __value = 0;
-  _CCCL_TRY_CUDA_API(::cudaDeviceGetAttribute, "failed to get device attribute", &__value, __attr, __dev);
-  return __value;
-}
-
-template <::cudaDeviceAttr _Attr>
-struct __dev_attr
```
Comment (bernhardmgruber) [inline cudax/include/cuda/experimental/__device/attributes.cuh:43]: Critical: this breaks with the forward declaration in device_ref:

https://github.com/NVIDIA/cccl/blob/7ff1d7bde11065ecd8328395e4e5e84191bc18d7/cudax/include/cuda/experimental/__device/device_ref.cuh#L36-L40

That specializes this template.

File: cudax/include/cuda/experimental/__device/attributes.cuh [hunk 2]
```diff
@@ -32,31 +32,8 @@ namespace cuda::experimental
...
-{
-  using type = int;
-
-  _CCCL_NODISCARD constexpr operator ::cudaDeviceAttr() const noexcept
-  {
-    return _Attr;
-  }
```
Comment (bernhardmgruber) [inline cudax/include/cuda/experimental/__device/attributes.cuh:50]: Important: it don't see this conversion anymore after your changs. This is a breaking change.

**Review by bernhardmgruber — CHANGES_REQUESTED** [review]
I think this change breaks the current design. Also, since this code was merged not too long ago, @ericniebler may have had reasons for splitting `__dev_attr` and `__dev_attr_with_type`. Let's hear what he says.

File: cudax/include/cuda/experimental/__device/attributes.cuh (hunk 1 above)
Comment (miscco) [reply cudax/include/cuda/experimental/__device/attributes.cuh:43]: We still define the same template:

```cpp
template <::cudaDeviceAttr _Attr>
struct __dev_attr : __dev_attr_impl<_Attr, int>
{};
```

File: cudax/include/cuda/experimental/__device/attributes.cuh (hunk 2 above)
Comment (miscco) [reply cudax/include/cuda/experimental/__device/attributes.cuh:50]: That conversion is still there on line 40, thankfully hidden by the github UI

Comment (bernhardmgruber) [reply cudax/include/cuda/experimental/__device/attributes.cuh:50]: :D

File: cudax/include/cuda/experimental/__device/attributes.cuh (hunk 1 above)
Comment (bernhardmgruber) [reply cudax/include/cuda/experimental/__device/attributes.cuh:43]: First PRs on Monday .. Thx!

**Review by bernhardmgruber — APPROVED** [review]
The change is fine, but I would still prefer to have @ericniebler's opinion on whether there was intent in splitting `__dev_attr` and `__dev_attr_with_type`.

**Review by ericniebler — APPROVED** [review]
looks good

**Comment by bernhardmgruber** [comment]
Great! Thanks for the contribution!

**Pull request merged**